File: frontend/src/lib/analysis.ts
import Papa from 'papaparse';
import * as XLSX from 'xlsx';

import { emotionOrder, type AnalysisResult, type BatchItem, type ColumnMapping, type DatasetRow, type EmotionName, type HistoryEntry } from '../types';

const STOPWORDS = new Set([
  'a',
  'an',
  'and',
  'are',
  'as',
  'at',
  'be',
  'but',
  'by',
  'for',
  'from',
  'has',
  'have',
  'i',
  'in',
  'is',
  'it',
  'its',
  'my',
  'of',
  'on',
  'or',
  'that',
  'the',
  'this',
  'to',
  'was',
  'we',
  'with',
  'you',
  'your'
]);

const JOY_WORDS = new Set(['amazing', 'awesome', 'brilliant', 'crisp', 'excellent', 'fantastic', 'great', 'impressive', 'love', 'perfect', 'reliable', 'smooth', 'stunning', 'wonderful', 'bright', 'fast', 'sleek']);
const ANGER_WORDS = new Set(['angry', 'annoying', 'bad', 'broken', 'buggy', 'defective', 'frustrating', 'horrible', 'laggy', 'poor', 'slow', 'terrible', 'weak']);
const SADNESS_WORDS = new Set(['disappointed', 'disappointing', 'dull', 'flat', 'mediocre', 'regret', 'sad', 'underwhelming', 'worse']);
const FEAR_WORDS = new Set(['afraid', 'anxious', 'concerned', 'fragile', 'risky', 'uncertain', 'worried']);
const SURPRISE_WORDS = new Set(['astonishing', 'unexpected', 'remarkable', 'surprising', 'shock', 'shocked', 'wow']);
const DISGUST_WORDS = new Set(['dirty', 'gross', 'nasty', 'offensive', 'repulsive', 'smelly', 'ugh']);

function tokenize(text: string): string[] {
  return text.toLowerCase().match(/[a-z0-9']+/g) ?? [];
}

function clamp(value: number, minimum: number, maximum: number): number {
  return Math.max(minimum, Math.min(maximum, value));
}

function normalizeCell(value: unknown): string {
  if (value === null || value === undefined) {
    return '';
  }
  return String(value).trim();
}

function scoreToken(token: string, scores: Record<EmotionName, number>): void {
  if (JOY_WORDS.has(token)) {
    scores.JOY += 1.8;
    scores.SURPRISE += 0.2;
  }
  if (ANGER_WORDS.has(token)) {
    scores.ANGER += 1.7;
  }
  if (SADNESS_WORDS.has(token)) {
    scores.SADNESS += 1.5;
  }
  if (FEAR_WORDS.has(token)) {
    scores.FEAR += 1.4;
  }
  if (SURPRISE_WORDS.has(token)) {
    scores.SURPRISE += 1.6;
  }
  if (DISGUST_WORDS.has(token)) {
    scores.DISGUST += 1.5;
  }
}

function buildRawEmotionScores(tokens: string[], customKeywords: string[] = []): Record<EmotionName, number> {
  const scores: Record<EmotionName, number> = {
    JOY: 1,
    ANGER: 0.75,
    SADNESS: 0.75,
    FEAR: 0.5,
    SURPRISE: 0.5,
    DISGUST: 0.5,
    NEUTRAL: 1.25
  };

  for (const token of tokens) {
    scoreToken(token, scores);
  }

  for (const keyword of customKeywords) {
    const normalizedKeyword = keyword.trim().toLowerCase();
    if (normalizedKeyword && tokens.includes(normalizedKeyword)) {
      scores.JOY += 0.8;
      scores.NEUTRAL += 0.15;
    }
  }

  return scores;
}

function normalizeEmotionScores(rawScores: Record<EmotionName, number>) {
  const total = Object.values(rawScores).reduce((sum, value) => sum + value, 0) || 1;
  return emotionOrder.map((emotion) => ({
    emotion,
    score: clamp(rawScores[emotion] / total, 0, 1)
  }));
}

function extractTags(tokens: string[], customKeywords: string[] = []): string[] {
  const tags: string[] = [];

  for (const token of tokens) {
    if (STOPWORDS.has(token) || token.length < 3 || tags.includes(token)) {
      continue;
    }
    tags.push(token);
    if (tags.length === 5) {
      return tags;
    }
  }

  for (const keyword of customKeywords) {
    const normalizedKeyword = keyword.trim().toLowerCase();
    if (!normalizedKeyword || tags.includes(normalizedKeyword)) {
      continue;
    }
    tags.push(normalizedKeyword);
    if (tags.length === 5) {
      return tags;
    }
  }

  return tags.length > 0 ? tags : ['review'];
}

function sentimentFromScores(rawScores: Record<EmotionName, number>): number {
  const positive = rawScores.JOY + rawScores.SURPRISE * 0.4;
  const negative = rawScores.ANGER + rawScores.SADNESS + rawScores.FEAR + rawScores.DISGUST;
  const denominator = Math.max(1, positive + negative);
  return Number(clamp((positive - negative) / denominator, -1, 1).toFixed(3));
}

function primaryEmotionFromScores(rawScores: Record<EmotionName, number>): EmotionName {
  return emotionOrder.reduce((bestEmotion, emotion) => (rawScores[emotion] > rawScores[bestEmotion] ? emotion : bestEmotion), 'NEUTRAL');
}

function confidenceFromScores(tokens: string[], rawScores: Record<EmotionName, number>): number {
  const dominant = Math.max(...Object.values(rawScores));
  const richness = Math.min(1, new Set(tokens).size / 18);
  const confidence = 0.55 + 0.35 * (dominant / Math.max(Object.values(rawScores).reduce((sum, value) => sum + value, 0), 1)) + 0.1 * richness;
  return Number(clamp(confidence, 0, 1).toFixed(3));
}

function buildSummary(emotion: EmotionName, sentiment: number, tags: string[]): string {
  const tone = sentiment > 0.15 ? 'Positive' : sentiment < -0.15 ? 'Negative' : 'Mixed';
  const focus = tags.slice(0, 3).join(', ') || 'the product';
  return `${tone} review with a ${emotion.toLowerCase()} signal, centered on ${focus}.`;
}

function cleanRow(row: Record<string, unknown>): DatasetRow {
  return Object.fromEntries(
    Object.entries(row).map(([key, value]) => [key, value === null || value === undefined ? null : String(value).trim() || null])
  );
}

export function mapHeadersToColumns(headers: string[]): ColumnMapping {
  const reviewColumn = bestMatch(headers, ['review', 'comment', 'feedback', 'opinion', 'content', 'text', 'message', 'remark']);
  const productName = bestMatch(headers, ['product name', 'item name', 'product title', 'product', 'item', 'device', 'device name', 'title']);
  const brand = bestMatch(headers, ['brand', 'manufacturer', 'maker', 'label']);
  const modelNumber = bestMatch(headers, ['model number', 'model no', 'model', 'sku', 'part number', 'serial number', 'number', 'no']);

  return {
    reviewColumn,
    productName,
    brand,
    modelNumber
  };
}

function normalizeHeader(header: string): string {
  return header.toLowerCase().replace(/[^a-z0-9]+/g, ' ').trim();
}

function bestMatch(headers: string[], keywords: string[], minimumScore = 2): string | null {
  let bestHeader: string | null = null;
  let bestScore = minimumScore - 1;

  for (const header of headers) {
    const normalized = normalizeHeader(header);
    let score = 0;

    for (const keyword of keywords) {
      if (normalized === keyword) {
        score += 4;
      } else if (normalized.includes(keyword)) {
        score += 2;
      } else if (normalized.split(' ').includes(keyword)) {
        score += 1;
      }
    }

    if (score > bestScore) {
      bestScore = score;
      bestHeader = header;
    }
  }

  return bestScore >= minimumScore ? bestHeader : null;
}

export function buildDemoAnalysisResult(input: { id: string; text: string; metadata?: Record<string, unknown> | null; customKeywords?: string[] }): AnalysisResult {
  const tokens = tokenize(input.text);
  const rawScores = buildRawEmotionScores(tokens, input.customKeywords ?? []);
  const emotions = normalizeEmotionScores(rawScores);
  const primaryEmotion = primaryEmotionFromScores(rawScores);
  const sentiment = sentimentFromScores(rawScores);
  const tags = extractTags(tokens, input.customKeywords ?? []);

  return {
    id: input.id,
    text: input.text,
    sentiment,
    primaryEmotion,
    emotions,
    tags,
    summary: buildSummary(primaryEmotion, sentiment, tags),
    confidenceScore: confidenceFromScores(tokens, rawScores),
    timestamp: Date.now(),
    metadata: input.metadata ?? null
  };
}

export async function parseDatasetFile(file: File): Promise<DatasetRow[]> {
  const fileName = file.name.toLowerCase();

  if (fileName.endsWith('.csv')) {
    const text = await file.text();
    const result = Papa.parse(text, {
      header: true,
      skipEmptyLines: true
    }) as { data?: Array<Record<string, unknown>>; errors: Array<{ message: string }> };

    if (result.errors.length > 0) {
      throw new Error(result.errors[0].message);
    }

    return (result.data ?? []).map((row: Record<string, unknown>) => cleanRow(row));
  }

  if (fileName.endsWith('.json')) {
    const text = await file.text();
    const payload = JSON.parse(text) as unknown;

    if (Array.isArray(payload)) {
      return payload.map((row) => cleanRow(row as Record<string, unknown>));
    }

    if (payload && typeof payload === 'object') {
      const objectPayload = payload as Record<string, unknown>;
      const rows = objectPayload.rows ?? objectPayload.items ?? objectPayload.data;
      if (Array.isArray(rows)) {
        return rows.map((row) => cleanRow(row as Record<string, unknown>));
      }
      return [cleanRow(objectPayload)];
    }

    throw new Error('Unsupported JSON dataset format');
  }

  if (fileName.endsWith('.xlsx') || fileName.endsWith('.xls')) {
    const buffer = await file.arrayBuffer();
    const workbook = XLSX.read(buffer, { type: 'array' });
    const sheetName = workbook.SheetNames[0];
    if (!sheetName) {
      return [];
    }

    const worksheet = workbook.Sheets[sheetName];
    const rows = XLSX.utils.sheet_to_json<Record<string, unknown>>(worksheet, { defval: null });
    return rows.map((row) => cleanRow(row));
  }

  throw new Error('Unsupported file type. Please upload CSV, JSON, or Excel files.');
}

export function buildMappedBatch(rows: DatasetRow[], mapping: ColumnMapping): BatchItem[] {
  const mapped: BatchItem[] = [];

  rows.forEach((row, index) => {
    const reviewColumn = mapping.reviewColumn ?? '';
    const reviewText = normalizeCell(reviewColumn ? row[reviewColumn] : row.review ?? row.text ?? '');

    if (!reviewText) {
      return;
    }

    const metadata: Record<string, unknown> = { ...row };
    if (mapping.reviewColumn) {
      delete metadata[mapping.reviewColumn];
    }

    mapped.push({
      id: normalizeCell((row.id ?? row.ID ?? row.Id) as unknown) || `row-${index + 1}`,
      text: reviewText,
      metadata
    });
  });

  return mapped;
}

export function buildHistoryEntry(params: {
  kind: 'single' | 'batch';
  title: string;
  results: AnalysisResult[];
  sourceName?: string;
}): HistoryEntry {
  const averageSentiment = params.results.length > 0 ? params.results.reduce((sum, result) => sum + result.sentiment, 0) / params.results.length : 0;
  const primaryEmotion = aggregatePrimaryEmotion(params.results);

  return {
    id: crypto.randomUUID(),
    kind: params.kind,
    title: params.title,
    sourceName: params.sourceName,
    results: params.results,
    primaryEmotion,
    averageSentiment: Number(averageSentiment.toFixed(3)),
    createdAt: Date.now()
  };
}


export function aggregatePrimaryEmotion(results: AnalysisResult[]): EmotionName {
  const scores = emotionOrder.reduce<Record<EmotionName, number>>(
    (accumulator, emotion) => ({ ...accumulator, [emotion]: 0 }),
    {} as Record<EmotionName, number>
  );

  for (const result of results) {
    scores[result.primaryEmotion] += 1;
  }

  return emotionOrder.reduce((bestEmotion, emotion) => (scores[emotion] > scores[bestEmotion] ? emotion : bestEmotion), 'NEUTRAL');
}

export function averageEmotionRadarData(results: AnalysisResult[]) {
  const aggregate = emotionOrder.reduce<Record<EmotionName, number>>(
    (accumulator, emotion) => ({ ...accumulator, [emotion]: 0 }),
    {} as Record<EmotionName, number>
  );

  if (results.length === 0) {
    return emotionOrder.map((emotion) => ({ emotion, score: 0 }));
  }

  for (const result of results) {
    for (const entry of result.emotions) {
      aggregate[entry.emotion] += entry.score;
    }
  }

  return emotionOrder.map((emotion) => ({ emotion, score: Number((aggregate[emotion] / results.length).toFixed(3)) }));
}

export function primaryEmotionDistribution(results: AnalysisResult[]) {
  const counts = emotionOrder.reduce<Record<EmotionName, number>>(
    (accumulator, emotion) => ({ ...accumulator, [emotion]: 0 }),
    {} as Record<EmotionName, number>
  );

  for (const result of results) {
    counts[result.primaryEmotion] += 1;
  }

  return emotionOrder.map((emotion) => ({ emotion, count: counts[emotion] }));
}

export function sentimentTimeline(results: AnalysisResult[]) {
  return results.map((result, index) => ({
    index: index + 1,
    sentiment: result.sentiment,
    label: result.id || `row-${index + 1}`
  }));
}

export function summarizeResults(results: AnalysisResult[]) {
  const averageSentiment = results.length > 0 ? results.reduce((sum, result) => sum + result.sentiment, 0) / results.length : 0;
  const averageConfidence = results.length > 0 ? results.reduce((sum, result) => sum + result.confidenceScore, 0) / results.length : 0;
  const primaryEmotion = aggregatePrimaryEmotion(results);

  return {
    total: results.length,
    averageSentiment: Number(averageSentiment.toFixed(3)),
    averageConfidence: Number(averageConfidence.toFixed(3)),
    primaryEmotion,
    positiveCount: results.filter((result) => result.sentiment > 0.15).length,
    negativeCount: results.filter((result) => result.sentiment < -0.15).length,
    neutralCount: results.filter((result) => Math.abs(result.sentiment) <= 0.15).length
  };
}

File: app/services/analysis.py
from __future__ import annotations

import asyncio
import inspect
import re
import time
from collections.abc import Awaitable, Callable
from typing import Any

from app.models.schemas import AnalysisResult, BatchItem, EmotionEnum, EmotionScore

import nltk
try:
    nltk.data.find('sentiment/vader_lexicon.zip')
except (LookupError, AttributeError):
    try:
        nltk.download('vader_lexicon', quiet=True)
    except Exception:
        pass

from nltk.sentiment.vader import SentimentIntensityAnalyzer
from textblob import TextBlob
from app.services.ml_classifier import SentixMLClassifier

try:
    vader_analyzer = SentimentIntensityAnalyzer()
except Exception:
    vader_analyzer = None

ml_classifier = SentixMLClassifier()



STOPWORDS = {
    "a",
    "an",
    "and",
    "are",
    "as",
    "at",
    "be",
    "but",
    "by",
    "for",
    "from",
    "has",
    "have",
    "i",
    "in",
    "is",
    "it",
    "its",
    "my",
    "of",
    "on",
    "or",
    "that",
    "the",
    "this",
    "to",
    "was",
    "we",
    "with",
    "you",
    "your",
}

JOY_WORDS = {
    "amazing", "awesome", "brilliant", "crisp", "excellent", "fantastic", "great", 
    "impressive", "love", "perfect", "reliable", "smooth", "stunning", "wonderful", 
    "bright", "fast", "sleek", "like", "likes", "liked", "good", "nice", "fine", 
    "happy", "pleased", "recommend", "recommended", "glad", "delighted", "loved", 
    "loves", "superb", "outstanding", "best", "perfectly", "well"
}
ANGER_WORDS = {
    "angry", "annoying", "bad", "broken", "buggy", "defective", "frustrating", 
    "horrible", "laggy", "poor", "slow", "terrible", "weak", "hate", "hated", 
    "hating", "dislike", "disliked", "annoyed", "mad", "worst"
}
SADNESS_WORDS = {
    "disappointed", "disappointing", "dull", "flat", "mediocre", "regret", "sad", 
    "underwhelming", "worse", "unhappy", "depressed", "sorry", "pity"
}
FEAR_WORDS = {"afraid", "anxious", "concerned", "fragile", "risky", "uncertain", "worried"}
SURPRISE_WORDS = {"astonishing", "unexpected", "remarkable", "surprising", "shock", "shocked", "wow"}
DISGUST_WORDS = {"dirty", "gross", "nasty", "offensive", "repulsive", "smelly", "ugh"}


def _tokenize(text: str) -> list[str]:
    return re.findall(r"[a-z0-9']+", text.lower())


def _clamp(value: float, minimum: float, maximum: float) -> float:
    return max(minimum, min(maximum, value))


def _build_raw_emotion_scores(tokens: list[str], custom_keywords: list[str] | None = None) -> dict[EmotionEnum, float]:
    # Baselines are balanced to ensure neutral sentiment outputs exactly 0.0:
    # positive = JOY (0.5) + 0.4 * SURPRISE (0.5) = 0.7
    # negative = ANGER (0.175) + SADNESS (0.175) + FEAR (0.175) + DISGUST (0.175) = 0.7
    # positive - negative = 0.0
    scores = {
        EmotionEnum.JOY: 0.5,
        EmotionEnum.ANGER: 0.175,
        EmotionEnum.SADNESS: 0.175,
        EmotionEnum.FEAR: 0.175,
        EmotionEnum.SURPRISE: 0.5,
        EmotionEnum.DISGUST: 0.175,
        EmotionEnum.NEUTRAL: 0.6,
    }

    for token in tokens:
        if token in JOY_WORDS:
            scores[EmotionEnum.JOY] += 1.8
            scores[EmotionEnum.SURPRISE] += 0.2
        if token in ANGER_WORDS:
            scores[EmotionEnum.ANGER] += 1.7
        if token in SADNESS_WORDS:
            scores[EmotionEnum.SADNESS] += 1.5
        if token in FEAR_WORDS:
            scores[EmotionEnum.FEAR] += 1.4
        if token in SURPRISE_WORDS:
            scores[EmotionEnum.SURPRISE] += 1.6
        if token in DISGUST_WORDS:
            scores[EmotionEnum.DISGUST] += 1.5

    for keyword in custom_keywords or []:
        normalized = keyword.lower().strip()
        if not normalized:
            continue
        if normalized in tokens:
            scores[EmotionEnum.JOY] += 0.8
            scores[EmotionEnum.NEUTRAL] += 0.15

    return scores


def _normalize_emotion_scores(raw_scores: dict[EmotionEnum, float]) -> list[EmotionScore]:
    total = sum(raw_scores.values()) or 1.0
    return [EmotionScore(emotion=emotion, score=_clamp(score / total, 0.0, 1.0)) for emotion, score in raw_scores.items()]


def _extract_tags(tokens: list[str], custom_keywords: list[str] | None = None) -> list[str]:
    tags: list[str] = []

    for token in tokens:
        if token in STOPWORDS or len(token) < 3:
            continue
        if token not in tags:
            tags.append(token)
        if len(tags) == 5:
            break

    for keyword in custom_keywords or []:
        normalized = keyword.lower().strip()
        if normalized and normalized not in tags:
            tags.append(normalized)
        if len(tags) == 5:
            break

    return tags or ["review"]


def _sentiment_from_scores(raw_scores: dict[EmotionEnum, float]) -> float:
    positive = raw_scores[EmotionEnum.JOY] + (0.4 * raw_scores[EmotionEnum.SURPRISE])
    negative = raw_scores[EmotionEnum.ANGER] + raw_scores[EmotionEnum.SADNESS] + raw_scores[EmotionEnum.FEAR] + raw_scores[EmotionEnum.DISGUST]
    denominator = max(1.0, positive + negative)
    sentiment = (positive - negative) / denominator
    return round(_clamp(sentiment, -1.0, 1.0), 3)


def _primary_emotion(raw_scores: dict[EmotionEnum, float]) -> EmotionEnum:
    return max(raw_scores, key=raw_scores.get)


def _confidence_score(tokens: list[str], raw_scores: dict[EmotionEnum, float]) -> float:
    dominant = max(raw_scores.values())
    richness = min(1.0, len(set(tokens)) / 18.0)
    confidence = 0.55 + (0.35 * dominant / max(sum(raw_scores.values()), 1.0)) + (0.1 * richness)
    return round(_clamp(confidence, 0.0, 1.0), 3)


def _summary(text: str, emotion: EmotionEnum, sentiment: float, tags: list[str]) -> str:
    tone = "positive" if sentiment > 0.15 else "negative" if sentiment < -0.15 else "mixed"
    focus = ", ".join(tags[:3]) if tags else "the product"
    return f"{tone.capitalize()} review with a {emotion.value.lower()} signal, centered on {focus}."


def build_demo_analysis_result(
    id: str,
    text: str,
    metadata: dict[str, Any] | None = None,
    custom_keywords: list[str] | None = None,
) -> AnalysisResult:
    tokens = _tokenize(text)
    raw_scores = _build_raw_emotion_scores(tokens, custom_keywords=custom_keywords)
    emotions = _normalize_emotion_scores(raw_scores)
    primary = _primary_emotion(raw_scores)
    rule_sentiment = _sentiment_from_scores(raw_scores)
    tags = _extract_tags(tokens, custom_keywords=custom_keywords)

    # Calculate VADER compound sentiment
    vader_compound = 0.0
    if vader_analyzer is not None:
        try:
            vader_scores = vader_analyzer.polarity_scores(text)
            vader_compound = float(vader_scores.get("compound", 0.0))
        except Exception:
            pass

    # Calculate Subjectivity using TextBlob
    subjectivity = 0.0
    try:
        blob = TextBlob(text)
        subjectivity = float(blob.sentiment.subjectivity)
    except Exception:
        pass

    # Calculate ML model sentiment prediction
    ml_sentiment = 0.0
    try:
        ml_sentiment = float(ml_classifier.predict(text))
    except Exception:
        pass

    # Blend them: VADER is highly reliable, ML sentiment is trained on synthetic corpus, and rule_sentiment is our keyword-based fallback.
    if vader_analyzer is not None:
        sentiment = 0.45 * vader_compound + 0.35 * ml_sentiment + 0.20 * rule_sentiment
    else:
        sentiment = 0.50 * ml_sentiment + 0.50 * rule_sentiment

    sentiment = round(_clamp(sentiment, -1.0, 1.0), 3)

    return AnalysisResult(
        id=id,
        text=text,
        sentiment=sentiment,
        primaryEmotion=primary,
        emotions=emotions,
        tags=tags,
        summary=_summary(text, primary, sentiment, tags),
        confidenceScore=_confidence_score(tokens, raw_scores),
        timestamp=int(time.time()),
        metadata=metadata,
        subjectivity=subjectivity,
        vaderCompound=vader_compound,
        mlSentiment=ml_sentiment,
    )


class SentixLLMService:
    async def analyze_text(self, text: str, metadata: dict[str, Any] | None = None) -> AnalysisResult:
        if not text.strip():
            raise ValueError("Text is required")

        # Run CPU-bound logic in a background thread to prevent event loop blocking
        return await asyncio.to_thread(
            build_demo_analysis_result,
            id="single-1",
            text=text,
            metadata=metadata,
            custom_keywords=None
        )

    async def analyze_batch(
        self,
        items: list[BatchItem],
        progress_callback: Callable[[int, int, AnalysisResult], Awaitable[None] | None] | None = None,
    ) -> list[AnalysisResult]:
        async def run_item(index: int, item: BatchItem) -> tuple[int, AnalysisResult]:
            result = await self.analyze_text(
                item.text,
                metadata=item.metadata | {"batchIndex": index},
            )
            result.id = item.id or f"row-{index + 1}"
            return index, result

        semaphore = asyncio.Semaphore(8)

        async def guarded_run(index: int, item: BatchItem) -> tuple[int, AnalysisResult]:
            async with semaphore:
                return await run_item(index, item)

        tasks = [asyncio.create_task(guarded_run(index, item)) for index, item in enumerate(items)]
        ordered_results: list[tuple[int, AnalysisResult]] = []
        completed = 0

        for future in asyncio.as_completed(tasks):
            index, result = await future
            ordered_results.append((index, result))
            completed += 1
            if progress_callback is not None:
                callback_result = progress_callback(completed, len(items), result)
                if inspect.isawaitable(callback_result):
                    await callback_result

        ordered_results.sort(key=lambda item: item[0])
        return [result for _, result in ordered_results]
